File: BeaconEmitter/BeaconEmitter/ViewController.m
//
//  ViewController.m
//  BeaconEmitter
//
//  Created by Oliver Drobnik on 20.05.14.
//  Copyright (c) 2014 Cocoanetics. All rights reserved.
//

#import "ViewController.h"

@interface ViewController () <CBPeripheralManagerDelegate>

@end

@implementation ViewController
{
   CBPeripheralManager *_peripheralManager;
   BOOL _isAdvertising;
}

- (void)viewDidLoad
{
   [super viewDidLoad];
   
   _peripheralManager =
   [[CBPeripheralManager alloc] initWithDelegate:self
                                           queue:nil];
   
   // changing autorization status kills app
   [self _updateUIForAuthorizationStatus];
}

- (void)_updateUIForAuthorizationStatus
{
   CBPeripheralManagerAuthorizationStatus
      status = [CBPeripheralManager authorizationStatus];
   
   if (status == CBPeripheralManagerAuthorizationStatusRestricted ||
       status == CBPeripheralManagerAuthorizationStatusDenied)
   {
      NSLog(@"Cannot start BT peripheral, not authorized");
      
      self.beaconSwitch.enabled = NO;
   }
   else
   {
      self.beaconSwitch.enabled = YES;
   }
}

- (void)_startAdvertising
{
   // get values from UI
   NSString *UUIDString = self.UUIDTextField.text;
   NSInteger major = [self.majorTextField.text integerValue];
   NSInteger minor = [self.minorTextField.text integerValue];
   NSUUID *UUID = [[NSUUID alloc] initWithUUIDString:UUIDString];
   
   // identifier cannot be nil, but is inconsequential
   CLBeaconRegion *region =
      [[CLBeaconRegion alloc] initWithProximityUUID:UUID
                                              major:major
                                              minor:minor
                                         identifier:@"FooBar"];
   
   // only care about this dictionary for advertising it
   NSDictionary *beaconPeripheralData =
      [region peripheralDataWithMeasuredPower:nil];
   [_peripheralManager startAdvertising:beaconPeripheralData];
}

- (void)_updateEmitterForDesiredState
{
   // only issue commands when powered on
   if (_peripheralManager.state == CBPeripheralManagerStatePoweredOn)
   {
      if (_isAdvertising)
      {
         if (!_peripheralManager.isAdvertising)
         {
            [self _startAdvertising];
         }
      }
      else
      {
         if (_peripheralManager.isAdvertising)
         {
            [_peripheralManager stopAdvertising];
         }
      }
   }
}

#pragma mark - CBPeripheralManagerDelegate

- (void)peripheralManagerDidUpdateState:(CBPeripheralManager *)peripheral
{
   [self _updateEmitterForDesiredState];
}

#pragma mark - Actions

- (IBAction)advertisingSwitch:(UISwitch *)sender
{
   _isAdvertising = sender.isOn;
   [self _updateEmitterForDesiredState];
}

@end

File: BeaconEmitter/Swift/ViewController.swift
//
//  ViewController.swift
//  BeaconEmitter
//
//  Created by Geoff Breemer on 11/10/14.
//  Copyright (c) 2014 Cocoanetics. All rights reserved.
//

import UIKit
import CoreBluetooth
import CoreLocation
import Foundation

@objc(ViewController) class ViewController: UIViewController, CBPeripheralManagerDelegate
{
    @IBOutlet var UUIDTextField: UITextField?
    @IBOutlet var majorTextField: UITextField?
    @IBOutlet var minorTextField: UITextField?
    
    @IBOutlet var beaconSwitch: UISwitch!
    
    private var _peripheralManager: CBPeripheralManager?
    private var _isAdvertising: Bool = false
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        _peripheralManager = CBPeripheralManager(delegate: self, queue: nil)
        
        // changing autorization status kills app
        _updateUIForAuthorizationStatus()
    }
    
    func _updateUIForAuthorizationStatus()
    {
        let status: CBPeripheralManagerAuthorizationStatus = CBPeripheralManager.authorizationStatus()
        
        if (status == .Restricted ||
            status == .Denied)
        {
            println("Cannot start BT peripheral, not authorized")
            
            beaconSwitch.enabled = false
        }
        else
        {
            beaconSwitch.enabled = true
        }
    }
    
    func _startAdvertising()
    {
        // get values from UI
        let UUIDString: String = UUIDTextField!.text
        let major: NSInteger = majorTextField!.text.toInt()!
        let minor: NSInteger = minorTextField!.text.toInt()!
        let UUID: NSUUID = NSUUID(UUIDString: UUIDString)!
        
        // identifier cannot be nil, but is inconsequential
        let region: CLBeaconRegion = CLBeaconRegion(proximityUUID: UUID, major: CLBeaconMajorValue(major), minor: CLBeaconMinorValue(minor), identifier: "FooBar")
        
        // only care about this dictionary for advertising it
        let beaconPeripheralData: Dictionary = region.peripheralDataWithMeasuredPower(nil)
        _peripheralManager!.startAdvertising(beaconPeripheralData)
    }
    
    func _updateEmitterForDesiredState()
    {
        // only issue commands when powered on
        if (_peripheralManager!.state == .PoweredOn)
        {
            if (_isAdvertising)
            {
                if (!_peripheralManager!.isAdvertising)
                {
                    _startAdvertising()
                }
            }
            else
            {
                if (_peripheralManager!.isAdvertising)
                {
                    _peripheralManager!.stopAdvertising()
                }
            }
        }
    }
    
    // MARK: - CBPeripheralManagerDelegate
    
    func peripheralManagerDidUpdateState(peripheral: CBPeripheralManager!) {
        _updateEmitterForDesiredState()
    }
    
    // MARK: - Actions
    
    @IBAction func advertisingSwitch(sender: UISwitch)
    {
        _isAdvertising = sender.on
        _updateEmitterForDesiredState()
    }
}